### ISSUE TITLE
any port for c#

### ISSUE THREAD

**Issue by radiaku** [dj-nitehawk/ThreadSafeAwesomium.Net#1]
Hello sir

Any port for c# browser.vb, i still no clue make awesomium thread working. 

Using awesomium 1.7.5

Thank you

**Comment by dj-nitehawk** [comment]
sorry mate, i'm not fluent in c#.

you can try using [sharp develop v4.4](http://www.icsharpcode.net/OpenSource/SD/Default.aspx) to convert this project to c#.

if your requirement is same as mine (to get the final rendered html output of a web page), you don't need to convert this code to c#. you can add this project to your visual studio solution and reference it from your c# project and simply use it.

Test.vb of this project shows how to use it. simply instantiate the browser class and call the `GetRenderedHTML` method.

**Comment by radiaku** [comment]
Wow, I didnt know sharp develop have convert ability. :) , previously I tried this http://converter.telerik.com/, but It have many error.

My goal is not just html, but automation and multiple scraping. This awesomium made my head spin off for 3 weeks. Maybe I should tried CEFsharp :)

**Comment by dj-nitehawk** [comment]
yes .net 4.5.

cefsharp is not as polished as awesomoum imho.

i haven't done any automation as of yet, so can't help there.

but i found a code snippet which should give u an idea: 

        void Html5_LoadingFrameComplete(object sender, Awesomium.Core.FrameEventArgs e)
        {
 
            // you can check url, do somthing before do this block.
 
            dynamic document = (Awesomium.Core.JSObject)Html5.ExecuteJavascriptWithResult("document");
 
            using (document)
            {
                var textbox = document.getElementsByTagName("input");
 
                if (textbox == null)
                    return;
 
                for (int i = 0; i < textbox.length; i++ )
                {
                    if (textbox[i].getAttribute("type") == "text")
                        textbox[i].value = "";
 
                    if (textbox[i].getAttribute("type") == "password")
                        textbox[i].value = "";
                }
            }
 
        }

basically what you do is handle the `LoadingFrameComplete` event and execute javascript in order to interact with the elements on page.

**Comment by radiaku** [comment]
thanks for the example. I already doing automation but when its coming  to running multiple step ( reusing 

For example
**Step 1**. I want to get element from first website ( screenshot ). 
**Step 2**. Then navigate to second website to post data. 

Here my code with error comment :) , go to button1_click  ( its too long to paste here )
http://pastebin.com/m6RNiBeL

if Iam using Cefsharp its much simpler and easier 
````
 static bool  ahb = true;
        private void button1_Click(object sender, EventArgs e)
        {
            button1.Enabled = false;

            const string testUrl = "http://www.mangahere.co/";


            // Create the offscreen Chromium browser.
            browser = new ChromiumWebBrowser(testUrl);

            // An event that is fired when the first page is finished loading.
            // This returns to us from another thread.
            browser.FrameLoadEnd += BrowserFrameLoadEnd;

            // We have to wait for something, otherwise the process will exit too soon.
            while (ahb)
            {
                Thread.Sleep(5000);

            }
            MessageBox.Show("Done step 1");

//========== here step2 ==========//

//========== end step2 ==========//


            // Clean up Chromium objects.  You need to call this in your application otherwise
            // you will get a crash when closing.
            Cef.Shutdown();

        }
````

**Comment by dj-nitehawk** [comment]
i made a little console app for you to show how to get a screenshot of a page.

get it from [here](https://mega.co.nz/#!aJsTXBAa!fLF3Q7WPFskxrLzZq68RTc4N03xMwJJSkNzPPwi4NRM)

it's in vb so you'll have to use a converter if you can't read vb.

you'll have to figure out the posting data to a form by yourself cause i dont have much free time these days. 

good luck!

**Comment by radiaku** [comment]
Thank mate for your effort. I still got error when webcore.dowork
````
 WebCore.DoWork(() =>
            {
                Session.Dispose();
                Session = null;
                return null;
            });
        }
````
error
````
Error	2	The type arguments for method 'Awesomium.Core.WebCore.DoWork<TResult>(System.Func<TResult>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.	C:\Users\flowingair\Documents\Visual Studio 2010\Projects\AccGodMode\AccGodMode\Class1.cs	40	13	AccGodMode
````
SS
![2015-05-06_012000](https://cloud.githubusercontent.com/assets/5983350/7479783/09bc3024-f38e-11e4-9d3b-7acd776a2b81.png)

I give up, I trying new solution using CEFSHARP 

thank for your time mate.

**Comment by radiaku** [comment]
Here finally without error convert vb to c#, only Invoke on  view.source didnt work.( I comment it ). I dont know why. View.LoadingFrameComplete never fired.
C# .net 45 , I leave here in case anyone maybe have better knowledge about C# than me. 
````

using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using Awesomium.Core;
using System.Threading.Tasks;
using System.Windows.Forms;

public class Class1 : IDisposable
{

    private static System.Threading.Thread Thread;
    private static bool CoreIsRunning = false;
    private static WebSession Session;

    WebView View;
    private static void AwesomiumThread()
    {
        if (WebCore.IsInitialized == false & CoreIsRunning == false)
        {
            WebCore.Initialize(new WebConfig { LogLevel = LogLevel.None }, false);
            WebCore.Run((s, e) => { 
                CoreIsRunning = true; 
            });
        }
    }

    public static void StartWebCore()
    {
        if (WebCore.IsInitialized == false && CoreIsRunning == false && (Thread == null))
        {
            Thread = new System.Threading.Thread(AwesomiumThread);
            Thread.Start();
        }
    }

    public static void StopWebCore()
    {
        if ((Session != null))
        {
            WebCore.DoWork(() =>
            {
                Session.Dispose();
                Session = null;
                return Session;
            });
        }
        WebCore.Shutdown();
        Thread = null;
        CoreIsRunning = false;
    }

    private static void InitSession()
    {
        if ((Session == null))
        {
            Session = WebCore.DoWork(() => { return WebCore.CreateWebSession(new WebPreferences { CanScriptsOpenWindows = false, DefaultEncoding = "utf-8", MaxHttpCacheStorage = 64000000 }); });
        }
    }

    private void CreateView()
    {
        if ((View != null))
        {
            View.LoadingFrameComplete -= null;
            WebCore.DoWork(() => 
            {
                if ((View != null))
                {
                    View.Dispose();
                }
                return View;
            });
            View = null;
        }
        View = WebCore.DoWork(() => { return WebCore.CreateWebView(1000, 1000, Session, WebViewType.Offscreen); });

    }

    public void Browser()
    {
        StartWebCore();
        InitSession();
        CreateView();
    }

    public void New()
    {
	    StartWebCore();
	    InitSession();
	    CreateView();
    }

    public void TakeScreenShot(string URL)
    {
        while (!(CoreIsRunning))
        {
            Task.Delay(100).Wait();
        }

        View.LoadingFrameComplete += (s, e) =>
        {
            if (e.IsMainFrame)
            {
                BitmapSurface surface = (BitmapSurface)View.Surface;
                surface.SaveToPNG("c:\\screenshot.png");
                //Interaction.MsgBox("Image saved!");
            }
        };


        //Here its doesnt WORK =======================//

        View.BeginInvoke(new Action(() => { View.Source = URL.ToUri(); }));

    }

    #region "IDisposable Support"
    private bool disposedValue;
    protected virtual void Dispose(bool disposing)
    {
        if (!this.disposedValue)
        {
            if (disposing)
            {
                View.LoadingFrameComplete -= null;
                WebCore.DoWork(() =>
                {
                    if ((View != null))
                    {
                        View.Dispose();
                        View = null;
                    }
                    return View;
                });
            }
        }
        this.disposedValue = true;
    }
    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }
    #endregion

}
````

**Comment by dj-nitehawk** [comment]
that's easy.

there's no BeginInvoke method on the view object.

change that line to: 

`View.Invoke(new Action(() => { View.Source = URL.ToUri(); }));`

also you might have to change the save path to something other than `c:\\` because i think dot net doesnt have permission to write to the C: drive.

**Comment by dj-nitehawk** [comment]
this works for me on my computer:

    using System;
    using System.Collections;
    using System.Collections.Generic;
    using System.Data;
    using System.Diagnostics;
    using Awesomium.Core;
    using System.Threading.Tasks;

    public class ScreenShot : IDisposable
    {

        private static System.Threading.Thread Thread;
        private static bool CoreIsRunning = false;
        private static WebSession Session;

        WebView View;
        private static void AwesomiumThread()
        {
            if (WebCore.IsInitialized == false & CoreIsRunning == false)
            {
                WebCore.Initialize(new WebConfig { LogLevel = LogLevel.None }, false);
                WebCore.Run((s, e) =>
                {
                    CoreIsRunning = true;
                });
            }
        }

        public static void StartWebCore()
        {
            if (WebCore.IsInitialized == false && CoreIsRunning == false && (Thread == null))
            {
                Thread = new System.Threading.Thread(AwesomiumThread);
                Thread.Start();
            }
        }

        public static void StopWebCore()
        {
            if ((Session != null))
            {
                WebCore.DoWork(() =>
                {
                    Session.Dispose();
                    Session = null;
                    return Session;
                });
            }
            WebCore.Shutdown();
            Thread = null;
            CoreIsRunning = false;
        }

        private static void InitSession()
        {
            if ((Session == null))
            {
                Session = WebCore.DoWork(() => { return WebCore.CreateWebSession(new WebPreferences { CanScriptsOpenWindows = false, DefaultEncoding = "utf-8", MaxHttpCacheStorage = 64000000 }); });
            }
        }

        private void CreateView()
        {
            if ((View != null))
            {
                View.LoadingFrameComplete -= null;
                WebCore.DoWork(() =>
                {
                    if ((View != null))
                    {
                        View.Dispose();
                    }
                    return View;
                });
                View = null;
            }
            View = WebCore.DoWork(() => { return WebCore.CreateWebView(1000, 1000, Session, WebViewType.Offscreen); });

        }


        public ScreenShot()
        {
            StartWebCore();
            InitSession();
            CreateView();
        }

        public void TakeScreenShot(string URL)
        {

            while (!(CoreIsRunning))
            {
                Task.Delay(100).Wait();
            }

            View.LoadingFrameComplete += (s, e) =>
            {
                if (e.IsMainFrame)
                {
                    BitmapSurface surface = (BitmapSurface)View.Surface;
                    surface.SaveToPNG("d:\\screenshot.png");
                    Console.WriteLine("Image saved!");
                }
            };


            View.Invoke(new Action(() => { View.Source = URL.ToUri(); }));

        }

        #region "IDisposable Support"
        private bool disposedValue;
        protected virtual void Dispose(bool disposing)
        {
            if (!this.disposedValue)
            {
                if (disposing)
                {
                    View.LoadingFrameComplete -= null;
                    WebCore.DoWork(() =>
                    {
                        if ((View != null))
                        {
                            View.Dispose();
                            View = null;
                        }
                        return View;
                    });
                }
            }
            this.disposedValue = true;
        }
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
        #endregion

    }

**Comment by radiaku** [comment]
Still dont work. Its this posible about GUI ( winform ) and console have different way ?

````
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using DeathByCaptcha;
using System.Net;
using System.IO;
using CsQuery;
using CsQuery.Web;
using System.Threading;
using Awesomium.Core;
using System.Threading.Tasks;
using RestSharp;
using System.Drawing.Imaging;
// Do not forget to reference DeathByCaptcha.dll in your project!
using DeathByCaptcha;
using Awesomium.Windows.Forms;
using Awesomium.Core.Data;

namespace AccGodMode
{
    public partial class MainAcc : Form
    {

        public MainAcc()
        {

            InitializeComponent();

            ScreenShot.StartWebCore();
        }
        static bool datate = true;

        private void button1_Click(object sender, EventArgs e)
        {
            button1.Enabled = false;

            using (ScreenShot browser = new ScreenShot())
            {
                browser.TakeScreenShot("http://google.com");

            }

            ScreenShot .StopWebCore();

        }


    }
}

````

**Comment by dj-nitehawk** [comment]
[here's a working forms project.](https://mega.co.nz/#!7AtkSDrL!R4oafFs7TqRPA9IiEMSUeNblTHNW9WIv1EkRpvYjrag)

enjoy!

**Comment by radiaku** [comment]
Wow that really worked out, Now I can run and reusing webview. 

Many thanks for your time, you're very kind. :)

Care to explain the  magic behind this...  ( what wrong with my code, I still have no clue  )

And where this little code called?
````
    public ScreenShot()
    {
        StartWebCore();
        InitSession();
        CreateView();
    }
````
I didnt really understand where this part called. :)


Please enlighten me?

**Comment by dj-nitehawk** [comment]
you're most welcome mate!

i just added this part: 

        while (!ScreenShotTaken)
        {
            Task.Delay(100).Wait();
        }

we have to wait until the `LoadingFrameComplete` is fired. otherwise the method will return before the event is fired and the `Screenshot` object gets disposed.

and `public ScreenShot() {...}` is called when you instantiate the Screenshot class with the following: 

`using (ScreenShot browser = new ScreenShot() ) {...}`

It's the c# constructor for the class.

good luck with your project...

**Comment by radiaku** [comment]
Great website mate, its iis server with vb.net technology behind?

Yeah, awesomium really pain in the ass with no proper documentation and newbie like me without C++ background. 

And thanks for the Enlightenment, sifu #bow

**Comment by dj-nitehawk** [comment]
thanks!
yeah it's a vb.net mvc app running on iis as the front-end and a winforms app running as the scheduler.
